Add book category logic and some package refactor:

DEz-1996 opened the pull request:
1) add category entity, controller, service, repository and Dtos
2) change book entity, controller, service, repository and Dtos with relation to book category
3) refactor dtos to corresponding folders

DEz-1996 committed:
Add CustomGlobalExceptionHandler class
Add validation for CreateBookRequestDto

DEz-1996 committed:
Some fixes at the student's advice

DEz-1996 committed:
Some fixes at the mentor's advice

DEz-1996 committed:
Add Pagination, Sorting, and Swagger to the existing controllers

DEz-1996 committed:
Add User entity with DTOs, mapper
Add AuthenticationController with register endpoint
Add FieldMatch annotation
Add RegistrationException and update GlobalExceptionHandler with it

DEz-1996 committed:
Some fixes with mentor's advice

DEz-1996 committed:
checkpoint

DEz-1996 committed:
Some fixes with mentor's advice

DEz-1996 committed:
Add Basic authentication
Add a Role entity and set access to the endpoints

DEz-1996 committed:
Some fixes with student's advices

DEz-1996 committed:
Implement JWT Security

DEz-1996 committed:
add task structure

DEz-1996 committed:
Add book category logic and some package refactor:
1) add category entity, controller, service, repository and Dtos
2) change book entity, controller, service, repository and Dtos with relation to book category
3) refactor dtos to corresponding folders

DEz-1996 committed:
Implement JWT Security (#11)

Implement JWT Security
* Merge branch 'main' into HW-11
# Conflicts:
delete UserServiceImp.java
delete FieldMatch.java
#	pom.xml
#	src/main/java/online/book/store/config/SecurityConfig.java
#	src/main/java/online/book/store/controller/AuthenticationController.java
#	src/main/java/online/book/store/controller/BookController.java
#	src/main/java/online/book/store/dto/CreateBookRequestDto.java
#	src/main/java/online/book/store/dto/UserRegistrationRequestDto.java
#	src/main/java/online/book/store/exception/CustomGlobalExceptionHandler.java
#	src/main/java/online/book/store/model/User.java
#	src/main/java/online/book/store/repository/UserRepository.java
#	src/main/java/online/book/store/repository/book/BookRepository.java
#	src/main/java/online/book/store/service/imp/BookServiceImpl.java
#	src/main/java/online/book/store/validation/FieldMatchValidator.java
#	src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml

DEz-1996 committed:
Minor fixes 1 (#13)

* Some minor fixes:
rename UserLoginResponceDto.java to UserLoginResponseDto.java
rename repository to roleRepository in UserServiceImpl.java
rename FieldMatchValidator.java to PasswordsFieldsMatchValidator.java
rename UserLoginResponceDto.java to UserLoginResponseDto.java
add @PreAuthorize and @Operation to searchBooks method in BookController.java

* Some minor fixes:
rename UserLoginResponceDto.java to UserLoginResponseDto.java
rename repository to roleRepository in UserServiceImpl.java
rename FieldMatchValidator.java to PasswordsFieldsMatchValidator.java
rename UserLoginResponceDto.java to UserLoginResponseDto.java
add @PreAuthorize and @Operation to searchBooks method in BookController.java

d1sam reviewed (changes requested):
why do I see changes related to **all** spring boot hws?

DEz-1996 committed:
Add book category logic and some package refactor:
1. add category entity, controller, service, repository and Dtos
2. change book entity, controller, service, repository and Dtos with relation to book category

DEz-1996 committed:
Fix checkstyle errors

vovasalyha commented on src/main/java/online/book/store/controller/CategoryController.java:
```diff
@@ -0,0 +1,96 @@
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PostMapping
+    @Operation(summary = "Create a new category", description = "Create a new category")
+    public CategoryDto createCategory(
```
```suggestion
    @ResponseStatus(HttpStatus.CREATED)
    public CategoryDto createCategory(
```

vovasalyha commented on src/main/java/online/book/store/model/Book.java:
```diff
@@ -32,4 +36,9 @@ public class Book {
+    @JoinTable(name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id"))
+    private Set<Category> categories;
```
Ignore it in equals, hashCode and toString.

vovasalyha commented on src/main/java/online/book/store/service/imp/CategoryServiceImp.java:
```diff
@@ -0,0 +1,60 @@
+    public CategoryDto update(CreateCategoryRequestDto categoryRequestDto, Long id) {
+        Category category = categoryMapper.toEntity(categoryRequestDto);
+        category.setId(id);
+        boolean isIdPresent = categoryRepository.findById(id).isPresent();
```
Better use existsById method.

vovasalyha reviewed (approved):
Approving, but see few minor comments.

DEz-1996 committed:
Some fixes:
1. Add annotation to ignore categories in equals, hashCode and toString in Book.java
2. Add @ResponseStatus(HttpStatus.CREATED) to createCategory method in CategoryController
3. Change validation logic in update method in CategoryServiceImp
